From a9760629678c792d9ade3c6ee5aee469ccba7a00 Mon Sep 17 00:00:00 2001
From: Seevalamuthu Mariappan <quic_seevalam@quicinc.com>
Date: Wed, 15 Jun 2022 19:11:45 +0530
Subject: [PATCH] ath12k: add hw_param for wakeup_mhi

Wakeup mhi is needed before pci_read/write only for QCA6390
and WCN6855. By default, wakeup & release mhi is enabled for all
hardwares
Hence, disable wakeup/release mhi using hw_param for other hardwares.

Found while discussing a similar fix for ath11k
https://lore.kernel.org/all/1636702019-26142-1-git-send-email-quic_seevalam@quicinc.com/

Signed-off-by: Seevalamuthu Mariappan <quic_seevalam@quicinc.com>
Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  1 +
 drivers/net/wireless/ath/ath12k/hw.h   |  1 +
 drivers/net/wireless/ath/ath12k/pci.c  | 20 ++++++++++++--------
 3 files changed, 14 insertions(+), 8 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/hw.h
+++ b/drivers/net/wireless/ath/ath12k/hw.h
@@ -196,6 +196,7 @@ struct ath12k_hw_params {
 
 	bool supports_ap_ps;
 	bool credit_flow;
+	bool wakeup_mhi;
 };
 
 /* BRINGUP: move to dp.h */
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1055,7 +1055,8 @@ u32 ath12k_pci_read32(struct ath12k_base
 	/* for offset beyond BAR + 4K - 32, may
 	 * need to wakeup MHI to access.
 	 */
-	if (test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
+	if (ab->hw_params->wakeup_mhi &&
+	    test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
 	    offset >= ACCESS_ALWAYS_OFF)
 		mhi_device_get_sync(ab_pci->mhi_ctrl->mhi_dev);
 
@@ -1099,7 +1100,8 @@ u32 ath12k_pci_read32(struct ath12k_base
 		}
 	}
 
-	if (test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
+	if (ab->hw_params->wakeup_mhi &&
+	    test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
 	    offset >= ACCESS_ALWAYS_OFF)
 		mhi_device_put(ab_pci->mhi_ctrl->mhi_dev);
 
@@ -1115,8 +1117,9 @@ void ath12k_pci_write32(struct ath12k_ba
 	/* for offset beyond BAR + 4K - 32, may
 	 * need to wakeup MHI to access.
 	 */
-	if (test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
-	    offset >= ACCESS_ALWAYS_OFF)
+	if (ab->hw_params->wakeup_mhi &&
+			test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
+			offset >= ACCESS_ALWAYS_OFF)
 		mhi_device_get_sync(ab_pci->mhi_ctrl->mhi_dev);
 
 	if (offset < WINDOW_START) {
@@ -1159,8 +1162,9 @@ void ath12k_pci_write32(struct ath12k_ba
 		}
 	}
 
-	if (test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
-	    offset >= ACCESS_ALWAYS_OFF)
+	if (ab->hw_params->wakeup_mhi &&
+			test_bit(ATH12K_PCI_FLAG_INIT_DONE, &ab_pci->flags) &&
+			offset >= ACCESS_ALWAYS_OFF)
 		mhi_device_put(ab_pci->mhi_ctrl->mhi_dev);
 }
 
--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -880,6 +880,7 @@ static const struct ath12k_hw_params ath
 
 		.supports_ap_ps = true,
 		.credit_flow = false,
+		.wakeup_mhi = false,
 	},
 	{
 		.name = "wcn7850 hw2.0",
@@ -944,6 +945,7 @@ static const struct ath12k_hw_params ath
 
 		.supports_ap_ps = true,
 		.credit_flow = false,
+		.wakeup_mhi = false,
 	},
 };
 
